### PR TITLE
feat(mesos-marathon): add marathon framework support for deis

### PR DESCRIPTION
**Pull request by smothiki**
Marathon framework integration with deis.
1. Install [deis](http://docs.deis.io/en/latest/) platform .
2. Follow the instructions to launch [controller]( http://docs.deis.io/en/latest/contributing/hacking/?highlight=hacking%20deis)
3. Build  deisctl 
4. ```deisctl install mesos``` and ```deisctl install marathon```
5. ```deisctl config controller set schedulerModule=mesos_marathon```

**Note :** 
*1. Right now deis is not using Marathon docker containerizers option. Executing a bash command to run docker containers.
2. If you are using Marathon deis limits CPU should be integer value of CPUs rather than CPU-Shares . By default  0.5 CPU is allocated to each APP.
3. Mesos Cluster is not HA*
**Future work:**
1. Move to Marathon docker containerizer and change publisher and logger logic accordingly to identify containers that are launched while creating an APP
2. Keep watching issue https://github.com/deis/deis/issues/3850 for design changes in deis to integrate schedulers.
3. Setup HA Mesos cluster using deisctl 

Refer: https://github.com/deis/deis/issues/3809
Requires #3692

### CODE REVIEW

**Comment by bacongobbler** [comment]
> 2. If you are using Marathon deis limits CPU should be integer value of CPUs rather than CPU-Shares . By default 0.5 CPU is allocated to each APP.

We should document this so users aren't confused. Would you mind writing documentation on all this, specifically:

 - how the architecture is laid out, both with internal and external mesos clusters
 - how users can migrate from fleet to mesos
 - how the mesos scheduler runs the containers
 - how can users connect to their external mesos cluster
 - a warning with reasoning behind why mesos is not HA

**Comment by smothiki** [comment]
OK. I'll keep in mind and write docs accordingly

File: deisctl/cmd/cmd.go [hunk 1]
```diff
@@ -181,6 +185,9 @@ func Stop(targets []string, b backend.Backend) error {
 		} else if targets[0] == swarm {
 			return StopSwarm(b)
 		}
+		if targets[0] == mesos {
```
Comment (carmstrong) [inline deisctl/cmd/cmd.go]: This should probably all be `else if` for consistency.

**Comment by technosophos** [comment]
I've reviewed the code. I'm not the best to judge the controller code, but everything I looked at seemed to be solid code. The deisctl code all looks great.

So consider this an LGTM on the code review.

**Comment by carmstrong** [comment]
> but we really should refactor all of these if/else if/else blocks into switch statements.

Ha! I was thinking the same thing when I implemented stateless-platform. For some reason, I thought Go doesn't support `switch`. I'm starting to get my languages mixed up...

File: controller/requirements.txt [hunk 2]
```diff
@@ -21,3 +21,4 @@ setproctitle==1.1.8
 static==1.1.1
 South==1.0.2
 python-ldap==2.4.19
+marathon
```
Comment (carmstrong) [inline controller/requirements.txt]: @smothiki We should probably lock this to a specific version of `marathon` so they don't break our code with updates.

**Comment by technosophos** [comment]
@carmstrong I might file an "easy fix" bug for that.

**Comment by carmstrong** [comment]
> I might file an "easy fix" bug for that.

SGTM. Great place for a new contributor to jump in.

File: controller/scheduler/mesos_marathon.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        """Run a one-off command"""
+        return self.fleet.run(name, image, entrypoint, command)
+
+    def state(self, name):
+        try:
+            for _ in xrange(30):
+                if self.client.get_app(self._app_id(name)).tasks_running == 1:
+                    return JobState.up
+                elif self.client.get_app(self._app_id(name)).tasks_running == 0:
+                    return JobState.created
```
Comment (bacongobbler) [inline controller/scheduler/mesos_marathon.py]: irrelevant to this PR, but I wonder if we should change this to `down` instead for more clarity. In all instances where we use `created`, the container is either initialized or down, which essentially mean the same thing.

File: docs/customizing_deis/choosing-a-scheduler.rst [hunk 4]
```diff
@@ -132,9 +131,67 @@ component, or spy on Docker events directly on the swarm-manager machine:
...
+The Marathon framework is now active. Commands such as ``deis destroy`` or
+``deis scale web=9`` will use `Marathon`_ to manage app containers.
+
+Deis starts Marathon on port 8180. You can manage the apps through Marathon UI which is accessible at http://<Marathon-node-IP>:8180
+
+.. note::
+
+    **Known Issues**
+
+    - deisctl install a standalone mesos cluster. As deis doesn't support metadata tags for fleet.
```
Comment (bacongobbler) [inline docs/customizing_deis/choosing-a-scheduler.rst]: What do you mean by this bullet point? I'm not sure what this is trying to explain

File: docs/customizing_deis/choosing-a-scheduler.rst [hunk 5]
```diff
@@ -132,9 +131,67 @@ component, or spy on Docker events directly on the swarm-manager machine:
...
+      keep watching `dynamic metadata fleet PR 1077`_.
+    - App containers will not be rescheduled if deis-registry is unavailable.
+    - Marathon starts on port 8180.
```
Comment (bacongobbler) [inline docs/customizing_deis/choosing-a-scheduler.rst]: how is this an issue? Does it need to be exposed in the ec2 security groups or something?

File: docs/customizing_deis/choosing-a-scheduler.rst [hunk 6]
```diff
@@ -132,9 +131,67 @@ component, or spy on Docker events directly on the swarm-manager machine:
...
+
+Deis starts Marathon on port 8180. You can manage the apps through Marathon UI which is accessible at http://<Marathon-node-IP>:8180
+
+.. note::
+
+    **Known Issues**
+
+    - deisctl install a standalone mesos cluster. As deis doesn't support metadata tags for fleet.
+      keep watching `dynamic metadata fleet PR 1077`_.
+    - App containers will not be rescheduled if deis-registry is unavailable.
```
Comment (bacongobbler) [inline docs/customizing_deis/choosing-a-scheduler.rst]: this is a known issue across all schedulers, no? If so we should remove this

File: controller/scheduler/mesos_marathon.py (hunk 3 above)
Comment (smothiki) [inline controller/scheduler/mesos_marathon.py]: SGTM

File: docs/customizing_deis/choosing-a-scheduler.rst (hunk 4 above)
Comment (smothiki) [inline docs/customizing_deis/choosing-a-scheduler.rst]: To install HA mesos cluster we have to specify metadata for each node. 
For example to install a HA mesos master we have to specify "metadata : master=true", then mesos master will get installed on all the nodes which have this metadata set. Right now deis deosn;t support metadata flags.

File: docs/customizing_deis/choosing-a-scheduler.rst (hunk 6 above)
Comment (smothiki) [inline docs/customizing_deis/choosing-a-scheduler.rst]: Not an issue for fleet. But a common problem for all other schedulers

File: docs/customizing_deis/choosing-a-scheduler.rst (hunk 4 above)
Comment (aledbf) [inline docs/customizing_deis/choosing-a-scheduler.rst]: @smothiki small correction: fleet doesn't support runtime change in the metadata. You can specify this in cloud-init during the deployment of the node.

Comment (smothiki) [inline docs/customizing_deis/choosing-a-scheduler.rst]: @aledbf yeah , thanks much clear message. I'll change accordingly.

File: docs/customizing_deis/choosing-a-scheduler.rst (hunk 5 above)
Comment (smothiki) [inline docs/customizing_deis/choosing-a-scheduler.rst]: Yes if users have to access Marathon UI.

File: docs/customizing_deis/choosing-a-scheduler.rst (hunk 4 above)
Comment (bacongobbler) [inline docs/customizing_deis/choosing-a-scheduler.rst]: Natively, deis does support metadata tags with the scheduler. The user has to set this up beforehand with fleet, then `deis tags:set` works as expected. The same can be done for `deisctl`, too. You just need to do some configuration with fleet by hand before it's booted to add `master=true` metadata to the node, which then the unit can utilize. I'm still not quite sure what the issue is here. :(

File: docs/customizing_deis/choosing-a-scheduler.rst (hunk 6 above)
Comment (bacongobbler) [inline docs/customizing_deis/choosing-a-scheduler.rst]: Would you mind linking this to the issue so users know where to go for more information?

File: docs/customizing_deis/choosing-a-scheduler.rst (hunk 4 above)
Comment (smothiki) [inline docs/customizing_deis/choosing-a-scheduler.rst]: Assuming a pre existing deis cluster which might not have metadata set for mesos cluster during cloud init.

File: docs/customizing_deis/choosing-a-scheduler.rst (hunk 5 above)
Comment (bacongobbler) [inline docs/customizing_deis/choosing-a-scheduler.rst]: Would you please change this to "If you want to access the Marathon UI, you'll have to expose port 8180 in the security group settings. This is blocked off by default for security purposes."

File: docs/customizing_deis/choosing-a-scheduler.rst [hunk 7]
```diff
@@ -132,9 +133,68 @@ component, or spy on Docker events directly on the swarm-manager machine:
...
+Deis starts Marathon on port 8180. You can manage the apps through Marathon UI which is accessible at http://<Marathon-node-IP>:8180
+
+.. note::
+
+    **Known Issues**
+
+    - deisctl install a standalone mesos cluster as fleet doesn't support runtime change in the metadata.
+      You can specify this in cloud-init during the deployment of the node. keep watching `dynamic metadata fleet PR 1077`_.
+    - If you want to access Marathon UI, you'll have to expose port 8180 in the security group settings.
+      This is blocked off by default for security purposes
```
Comment (bacongobbler) [inline docs/customizing_deis/choosing-a-scheduler.rst]: missing a period at the end for consistency

File: docs/customizing_deis/choosing-a-scheduler.rst [hunk 8]
```diff
@@ -132,9 +133,68 @@ component, or spy on Docker events directly on the swarm-manager machine:
...
+The Marathon framework is now active. Commands such as ``deis destroy`` or
+``deis scale web=9`` will use `Marathon`_ to manage app containers.
+
+Deis starts Marathon on port 8180. You can manage the apps through Marathon UI which is accessible at http://<Marathon-node-IP>:8180
+
+.. note::
+
+    **Known Issues**
+
+    - deisctl install a standalone mesos cluster as fleet doesn't support runtime change in the metadata.
```
Comment (bacongobbler) [inline docs/customizing_deis/choosing-a-scheduler.rst]: installs*

File: docs/customizing_deis/choosing-a-scheduler.rst [hunk 9]
```diff
@@ -132,9 +133,68 @@ component, or spy on Docker events directly on the swarm-manager machine:
...
+    - If you want to access Marathon UI, you'll have to expose port 8180 in the security group settings.
+      This is blocked off by default for security purposes
+    - Deis does not yet use Marathon's docker container API to create containers.
+    - CPU shares are integer values of number of CPUs and Memory limits unit should be specified in MB.
+
+
 .. _Kubernetes: http://kubernetes.io/
-.. _Mesos: http://mesos.apache.org/
+.. _Mesos: http://mesos.apache.org
+.. _Marathon: https://github.com/mesosphere/marathon#marathon-
```
Comment (bacongobbler) [inline docs/customizing_deis/choosing-a-scheduler.rst]: the `#marathon-` at the end doesn't link to anything as far as I'm aware. We should just shorten this to https://github.com/mesosphere/marathon.

**Comment by phspagiari** [comment]
Yey! Mesos on Deis its awesome. I will try to help using on my environment. 
But guys... where is the tests? Its missing the label "need tests", right?

**Comment by smothiki** [comment]
@phspagiari This is a release tech preview. So no tests for now.

**Comment by carmstrong** [comment]
Code LGTM!

**Comment by bacongobbler** [comment]
This is a preview so bugs are to be expected. We aren't promoting this for production use so we don't want to add this to the integration test suite, however we still perform manual "happy path" tests to make sure that it's shippable at the bare minimum :)

**Comment by lorieri** [comment]
Can I make a tech preview too ?

**Comment by carmstrong** [comment]
> Can I make a tech preview too ?

We use the term "tech preview" for our scheduler integrations, because many of them are not yet production-ready. PRs for anything you'd like to contribute are great! What are you thinking, @lorieri?

**Comment by lorieri** [comment]
I'm jealous how easily it got a 'LGTM' from you...

**Comment by carmstrong** [comment]
> I'm jealous how easily it got a 'LGTM' from you...

@smothiki and @kmala work for us full-time ;) There's been a lot of discussion and coordination behind-the-scenes.

**Comment by lorieri** [comment]
Cool :)

**Comment by carmstrong** [comment]
Latest test run: https://ci.deis.io/job/test-ec2/1714/

File: docs/customizing_deis/choosing-a-scheduler.rst [hunk 10]
```diff
@@ -132,9 +134,68 @@ component, or spy on Docker events directly on the swarm-manager machine:
...
+Then set the controller's ``schedulerModule`` to "mesos_marathon":
+
+.. code-block:: console
+
+    $ deisctl config controller set schedulerModule=mesos_marathon
+
+The Marathon framework is now active. Commands such as ``deis destroy`` or
+``deis scale web=9`` will use `Marathon`_ to manage app containers.
+
+Deis starts Marathon on port 8180. You can manage the apps through Marathon UI which is accessible at http://<Marathon-node-IP>:8180
```
Comment (mboersma) [inline docs/customizing_deis/choosing-a-scheduler.rst]: I would move "the" later and add a comma:
> You can manage apps through the Marathon UI, which is accessible at...

File: docs/customizing_deis/choosing-a-scheduler.rst [hunk 11]
```diff
@@ -132,9 +134,68 @@ component, or spy on Docker events directly on the swarm-manager machine:
...
+
+.. note::
+
+    **Known Issues**
+
+    - deisctl installs a standalone mesos cluster as fleet doesn't support runtime change in the metadata.
```
Comment (mboersma) [inline docs/customizing_deis/choosing-a-scheduler.rst]: "... as fleet doesn't support runtime changes to metadata."

File: docs/customizing_deis/choosing-a-scheduler.rst [hunk 12]
```diff
@@ -132,9 +134,68 @@ component, or spy on Docker events directly on the swarm-manager machine:
...
+      You can specify this in cloud-init during the deployment of the node. keep watching `dynamic metadata fleet PR 1077`_.
+    - If you want to access Marathon UI, you'll have to expose port 8180 in the security group settings.
+      This is blocked off by default for security purposes.
+    - Deis does not yet use Marathon's docker container API to create containers.
+    - CPU shares are integer values of number of CPUs and Memory limits unit should be specified in MB.
```
Comment (mboersma) [inline docs/customizing_deis/choosing-a-scheduler.rst]: "CPU shares are integers representing the number of CPUs. Memory limits should be specified in MB."

File: docs/customizing_deis/choosing-a-scheduler.rst [hunk 13]
```diff
@@ -132,9 +134,68 @@ component, or spy on Docker events directly on the swarm-manager machine:
...
+``deis scale web=9`` will use `Marathon`_ to manage app containers.
+
+Deis starts Marathon on port 8180. You can manage the apps through Marathon UI which is accessible at http://<Marathon-node-IP>:8180
+
+.. note::
+
+    **Known Issues**
+
+    - deisctl installs a standalone mesos cluster as fleet doesn't support runtime change in the metadata.
+      You can specify this in cloud-init during the deployment of the node. keep watching `dynamic metadata fleet PR 1077`_.
```
Comment (mboersma) [inline docs/customizing_deis/choosing-a-scheduler.rst]: The last sentence should be capitalized and perhaps expanded:
```
Keep watching `dynamic metadata fleet PR 1077`_ for support of runtime changes to metadata.
```

**Comment by mboersma** [comment]
So where is the code for the various mesos-* and zookeeper containers? I assumed this would be similar to swarm, where the containers are part of `make build`, but I can't tell from this PR what the origin of those containers is. The Mesos / Marathon containers need to be versioned with Deis as part of the build / CI process.

**Comment by smothiki** [comment]
changed marathon python client version to 0.6.15 as version 0.7.0 has issues with names having more than two dots.

File: controller/scheduler/mesos_marathon.py [hunk 14]
```diff
@@ -0,0 +1,111 @@
+import re
+import time
+from django.conf import settings
+from marathon import MarathonClient
+from marathon.models import MarathonApp
+from .states import JobState
+from docker import Client
+from .fleet import FleetHTTPClient
```
Comment (mboersma) [inline controller/scheduler/mesos_marathon.py]: These imports should have a blank line between each group as per [PEP8](https://www.python.org/dev/peps/pep-0008/#imports):

> Imports should be grouped in the following order:
> - standard library imports
> - related third party imports
> - local application/library specific imports
>
> *You should put a blank line between each group of imports.*

File: controller/scheduler/mesos_marathon.py [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+        if mems:
+            mems = mems.lower()
+            if mems[-2:-1].isalpha() and mems[-1].isalpha():
+                mems = mems[:-1]
+            m = int(mems[:-1])
+        c = 0.5
+        cpu = kwargs.get('cpu', {}).get(l['c_type'])
+        if cpu:
+            c = cpu
+        cpu = kwargs.get('cpu', {}).get(l['c_type'])
```
Comment (mboersma) [inline controller/scheduler/mesos_marathon.py]: This line is the same as line 47 and appears to be a no-op.

Comment (smothiki) [inline controller/scheduler/mesos_marathon.py]: sorry my bad. I will remove 47

File: controller/scheduler/mesos_marathon.py [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+        self.client.create_app(app_id,
+                               MarathonApp(cmd="docker run --name "+name+" -P "+image+" "+command,
+                                           mem=m, cpus=c))
```
Comment (mboersma) [inline controller/scheduler/mesos_marathon.py]: This is a bit hard to read, and generally speaking string formatting is preferable to concatenation in python to avoid object copies. I would do this:
```python
marathon_app = MarathonApp("docker run --name {name} -P {image} {command}".format(**locals), mem=m, cpus=c)
self.client.create_app(app_id, marathon_app)
```

File: controller/scheduler/mesos_marathon.py [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+        self.client.scale_app(app_id, 0, force=True)
+        for _ in xrange(30):
```
Comment (mboersma) [inline controller/scheduler/mesos_marathon.py]: If 30 is an arbitrary number used elsewhere, let's make it a module-level constant instead like `POLL_ATTEMPTS = 30`.

File: controller/scheduler/mesos_marathon.py [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+            self._delete_container(host, name)
+        except:
+            self.client.delete_app(self._app_id(name), force=True)
+
+    def _delete_container(self, host, name):
+        docker_cli = Client("tcp://{}:2375".format(host), timeout=1200, version='1.17')
+        try:
+            if docker_cli.inspect_container(name)['State']:
+                docker_cli.remove_container(name, force=True)
+        except:
```
Comment (mboersma) [inline controller/scheduler/mesos_marathon.py]: A [naked `except`](http://www.wilfred.me.uk/blog/2013/11/03/no-naked-excepts/) that doesn't also `raise` the Exception it caught can only lose information and submerge actual bugs. Always declare what type of Exception you expect to encounter. The python interpreter itself could throw a [`MemoryError`](https://docs.python.org/2/library/exceptions.html#exceptions.MemoryError) or [`SystemError`](https://docs.python.org/2/library/exceptions.html#exceptions.SystemError) at any time, and this code says "I don't care and I want to hide that."

Really `except` should catch a specific Exception type, or types, with only very rare...um...exceptions. :

File: controller/scheduler/mesos_marathon.py [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+                    return JobState.created
+                time.sleep(1)
+        except:
+            return JobState.destroyed
+
+    def attach(self, name):
+        """
+        Attach to a job's stdin, stdout and stderr
+        """
+        raise NotImplementedError
```
Comment (mboersma) [inline controller/scheduler/mesos_marathon.py]: This no-op is the same in fleet.py, and it makes me think we're at the point of declaring a common `Scheduler` superclass for `MarathonHTTPClient`, `FleetHTTPClient`, and `SwarmClient` (and Mock/Chaos/K8S too). Then it defines the common interface explicitly and raises `NotImplementedError` in its default implementations.

File: deisctl/units/deis-mesos-marathon.service [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+EnvironmentFile=/etc/environment
+Restart=on-failure
+RestartSec=20
+TimeoutStartSec=0
+ExecStartPre=-/bin/sh -c "etcdctl get /deis/scheduler/mesos/marathon >/dev/null 2>&1 || etcdctl mk /deis/scheduler/mesos/marathon"
+ExecStartPre=/bin/sh -c "etcdctl set /deis/scheduler/mesos/marathon $COREOS_PRIVATE_IPV4"
+ExecStartPre=/bin/sh -c "IMAGE=`/run/deis/bin/get_image /deis/mesos-marathon` && docker history $IMAGE >/dev/null 2>&1 || docker pull $IMAGE"
+ExecStartPre=-/usr/bin/docker kill deis-mesos-marathon
+ExecStartPre=-/usr/bin/docker rm deis-mesos-marathon
+ExecStart=/usr/bin/sh -c "IMAGE=`/run/deis/bin/get_image /deis/mesos-marathon` && docker run \
```
Comment (mboersma) [inline deisctl/units/deis-mesos-marathon.service]: Could this be a one-liner? I know it's long, but we don't break onto multiple lines in other unit files.

File: docs/customizing_deis/choosing-a-scheduler.rst [hunk 21]
```diff
@@ -97,10 +103,7 @@ to `enable the remote API`_.
     **Known Issues**
 
     - It is not yet possible to change the default affinity filter.
-    - If swarm can't create all the containers requested, Deis returns an error
-      and leaves some containers in the "created" state. Until this behavior is fixed,
-      be mindful of resource limitations on your cluster.
-    - App containers will not be rescheduled if deis-registry is unavailable.
```
Comment (mboersma) [inline docs/customizing_deis/choosing-a-scheduler.rst]: Did you mean to remove this bullet point about deis-registry? I thought that was still an issue...

File: docs/customizing_deis/choosing-a-scheduler.rst [hunk 22]
```diff
@@ -97,10 +103,7 @@ to `enable the remote API`_.
...
+    - If swarm can't create all the containers requested during scale, deis rolls back the scale operation.
```
Comment (mboersma) [inline docs/customizing_deis/choosing-a-scheduler.rst]: I think now that you fixed `deis scale` to rollback if needed, this isn't a "Known Issue" any more, but the correct behavior. So maybe just delete the original bullet point but don't add this one in its place.

Comment (smothiki) [inline docs/customizing_deis/choosing-a-scheduler.rst]: Sure I will remove the bullet point.

File: docs/customizing_deis/choosing-a-scheduler.rst (hunk 21 above)
Comment (smothiki) [inline docs/customizing_deis/choosing-a-scheduler.rst]: Registry is a common issue so I added it as note at the top of the doc line 17

**Comment by smothiki** [comment]
updated code as per @mboersma's comments

File: controller/scheduler/mesos_marathon.py [hunk 23]
```diff
@@ -0,0 +1,109 @@
...
+            mems = mems.lower()
+            if mems[-2:-1].isalpha() and mems[-1].isalpha():
+                mems = mems[:-1]
+            m = int(mems[:-1])
+        c = 0.5
+        cpu = kwargs.get('cpu', {}).get(l['c_type'])
+        if cpu:
+            c = cpu
+        cmd = "docker run --name {name} -P {image} {command}".format(**locals())
+        self.client.create_app(app_id,MarathonApp(cmd=cmd, mem=m, cpus=c))
```
Comment (mboersma) [inline controller/scheduler/mesos_marathon.py]: Needs a space after `app_id,` or it will fail `flake8` checks.

File: docs/customizing_deis/choosing-a-scheduler.rst (hunk 21 above)
Comment (mboersma) [inline docs/customizing_deis/choosing-a-scheduler.rst]: +1

**Comment by mboersma** [comment]
I tried to build the mesos-* and zookeeper-* images with `make -C mesos/ build`, but it appears we need to vendor in `go-bindata` with `godep`:
```console
$ make -C mesos/ build
...
Building...
go-bindata -pkg bindata -o pkg/boot/zookeeper/bindata/bindata.go pkg/boot/zookeeper/bash/; \
	GOOS=linux GOARCH=amd64 CGO_ENABLED=0 godep go build -a -installsuffix cgo -ldflags '-s' -o zookeeper/bin/boot pkg/boot/zookeeper/main/boot.go
/bin/sh: go-bindata: command not found
```

**Comment by mboersma** [comment]
When building, `go vet` also reports a lot of warnings about missing or malformed comments that should be cleaned up before this is merged. Also since the `extpoints` files are generated, we should exclude them from `make test-style` target.
```console
$ make -C mesos/ build
godep go vet github.com/deis/deis/mesos/pkg/boot github.com/deis/deis/mesos/pkg/confd github.com/deis/deis/mesos/pkg/etcd github.com/deis/deis/mesos/pkg/fleet github.com/deis/deis/mesos/pkg/log github.com/deis/deis/mesos/pkg/net
pkg/boot/main.go:17:2: a blank import should be only in a main or test package, or have a comment justifying it
pkg/boot/extpoints/extpoints.go:1:1: package comment should be of the form "Package extpoints ..."
pkg/boot/extpoints/extpoints.go:34:1: exported function RegisterExtension should have comment or be unexported
pkg/boot/extpoints/extpoints.go:46:1: exported function UnregisterExtension should have comment or be unexported
pkg/boot/extpoints/extpoints.go:131:5: exported var BootComponents should have comment or be unexported
pkg/boot/mesos/marathon/main.go:32:6: exported type MesosBoot should have comment or be unexported
pkg/boot/mesos/marathon/main.go:34:1: exported method MesosBoot.MkdirsEtcd should have comment or be unexported
pkg/boot/mesos/marathon/main.go:38:1: exported method MesosBoot.EtcdDefaults should have comment or be unexported
pkg/boot/mesos/marathon/main.go:42:1: exported method MesosBoot.PreBootScripts should have comment or be unexported
pkg/boot/mesos/marathon/main.go:53:1: exported method MesosBoot.PreBoot should have comment or be unexported
pkg/boot/mesos/marathon/main.go:57:1: exported method MesosBoot.BootDaemons should have comment or be unexported
pkg/boot/mesos/marathon/main.go:63:1: exported method MesosBoot.WaitForPorts should have comment or be unexported
pkg/boot/mesos/marathon/main.go:67:1: exported method MesosBoot.PostBootScripts should have comment or be unexported
pkg/boot/mesos/marathon/main.go:71:1: exported method MesosBoot.PostBoot should have comment or be unexported
pkg/boot/mesos/marathon/main.go:75:1: exported method MesosBoot.ScheduleTasks should have comment or be unexported
pkg/boot/mesos/marathon/main.go:79:1: exported method MesosBoot.UseConfd should have comment or be unexported
pkg/boot/mesos/marathon/main.go:83:1: exported method MesosBoot.PreShutdownScripts should have comment or be unexported
pkg/boot/mesos/master/main.go:30:6: exported type MesosBoot should have comment or be unexported
pkg/boot/mesos/master/main.go:32:1: exported method MesosBoot.MkdirsEtcd should have comment or be unexported
pkg/boot/mesos/master/main.go:36:1: exported method MesosBoot.EtcdDefaults should have comment or be unexported
pkg/boot/mesos/master/main.go:40:1: exported method MesosBoot.PreBootScripts should have comment or be unexported
pkg/boot/mesos/master/main.go:44:1: exported method MesosBoot.PreBoot should have comment or be unexported
pkg/boot/mesos/master/main.go:48:1: exported method MesosBoot.BootDaemons should have comment or be unexported
pkg/boot/mesos/master/main.go:56:1: exported method MesosBoot.WaitForPorts should have comment or be unexported
pkg/boot/mesos/master/main.go:60:1: exported method MesosBoot.PostBootScripts should have comment or be unexported
pkg/boot/mesos/master/main.go:64:1: exported method MesosBoot.PostBoot should have comment or be unexported
pkg/boot/mesos/master/main.go:68:1: exported method MesosBoot.ScheduleTasks should have comment or be unexported
pkg/boot/mesos/master/main.go:72:1: exported method MesosBoot.UseConfd should have comment or be unexported
pkg/boot/mesos/master/main.go:76:1: exported method MesosBoot.PreShutdownScripts should have comment or be unexported
pkg/boot/mesos/slave/main.go:30:6: exported type MesosBoot should have comment or be unexported
pkg/boot/mesos/slave/main.go:32:1: exported method MesosBoot.MkdirsEtcd should have comment or be unexported
pkg/boot/mesos/slave/main.go:36:1: exported method MesosBoot.EtcdDefaults should have comment or be unexported
pkg/boot/mesos/slave/main.go:40:1: exported method MesosBoot.PreBootScripts should have comment or be unexported
pkg/boot/mesos/slave/main.go:44:1: exported method MesosBoot.PreBoot should have comment or be unexported
pkg/boot/mesos/slave/main.go:48:1: exported method MesosBoot.BootDaemons should have comment or be unexported
pkg/boot/mesos/slave/main.go:55:1: exported method MesosBoot.WaitForPorts should have comment or be unexported
pkg/boot/mesos/slave/main.go:59:1: exported method MesosBoot.PostBootScripts should have comment or be unexported
pkg/boot/mesos/slave/main.go:63:1: exported method MesosBoot.PostBoot should have comment or be unexported
pkg/boot/mesos/slave/main.go:67:1: exported method MesosBoot.ScheduleTasks should have comment or be unexported
pkg/boot/mesos/slave/main.go:71:1: exported method MesosBoot.UseConfd should have comment or be unexported
pkg/boot/mesos/slave/main.go:75:1: exported method MesosBoot.PreShutdownScripts should have comment or be unexported
pkg/boot/zookeeper/zookeeper_server.go:9:6: exported type ZkServer should have comment or be unexported
```

**Comment by mboersma** [comment]
(I can't find any clean way to exclude just the `extpoints` package from the linter tools, so @smothiki and I agreed just to ignore those couple of warnings.)

**Comment by mboersma** [comment]
This PR should add some build artifact directories to the root .gitignore so they aren't committed to version control:
```
mesos/**/bin/
mesos/**/bindata/
```

**Comment by mboersma** [comment]
More comments:
- the git commit subjects should refer consistently to `feat(mesos)` or `docs(mesos)`, etc.
- the two `fix(style)` commits should have been `style(mesos)`, but instead should probably be squashed into the `feat(mesos)` commits so the PR is a clean unit of work
- it needs a `make deploy` target like swarm/ and the other components have, which really helps with testing
- could we somehow make @aledbf the author of 57bb2e7e70451ecd2bfd1d11976b2f926bf078ee...? He deserves so much credit here. :clap:

**Comment by smothiki** [comment]
Every container occupies 0.5 of the total Cpus by default. So scaling beyond cluster resources will result in errors.

**Comment by smothiki** [comment]
Using Marathon framework we are running docker run command to start apps in docker containers. Marathon will return once the command is executed but docker will take time to pull the image from registry and start container. Fixed the function return time from create container , ```git push  ``` exits  only if the container is created.

**Comment by mboersma** [comment]
This should have a mesos/README.md. See #4067 for an example.

**Comment by smothiki** [comment]
Added README.md

**Comment by mboersma** [comment]
I think the commit messages should reference "mesos", not "mesos-marathon" since the scope of the changes is limited to the `mesos/` directory.

I can't run `make -C mesos/ deploy`, it always fails trying to tag an image:
```
Removing intermediate container 574eba359c3d
Step 8 : ENTRYPOINT /app/bin/boot
 ---> Running in 78f5b724074c
 ---> 5150a3d8a710
Removing intermediate container 78f5b724074c
Successfully built 5150a3d8a710
docker tag -f deisci/mesos-master:git-4b11f51 192.168.59.103:5000/deisci/mesos-master:git-4b11f51
Error response from daemon: could not find image: no such id: deisci/mesos-master:git-4b11f51
make: *** [push] Error 1
```

**Comment by smothiki** [comment]
Image prefix is hard-coded to deis/ . Fixed Makefile to read from Variable $(IMAGE_PREFIX)

File: mesos/README.md [hunk 24]
```diff
@@ -0,0 +1,23 @@
+# Mesos with Marathon Framework
+
+[Marathon Framework](https://github.com/mesosphere/marathon) components for use as
+an optional scheduler in Deis, the open source PaaS.
+
+## Usage
+
+Please consult the [Mesos with Marathon framework](http://docs.deis.io/en/latest/customizing_deis/choosing-a-scheduler/#Mesos-with-Marathon-framework)
```
Comment (mboersma) [inline mesos/README.md]: Anchor links are actually treated as case-sensitive, so this link should be changed to lower-case or it will just bring you to the top of the web page:
`http://docs.deis.io/customizing_deis/choosing-a-scheduler/#mesos-with-marathon-framework`

**Comment by mboersma** [comment]
Does `make -C mesos/ deploy` work for you? It's still failing for me after the most recent changes:
```
Successfully built 0660b059be15
rm -f Dockerfile
cp marathon Dockerfile
docker cp `docker create deis/build-mesos-marathon:git-6c28430 /bin/bash`:/marathon/target/marathon-assembly-0.8.2-RC3.jar .
Unable to find image 'deis/build-mesos-marathon:git-6c28430' locally
Pulling repository deis/build-mesos-marathon
Error: image deis/build-mesos-marathon:git-6c28430 not found
mv marathon-assembly-0.8.2-RC3.jar marathon-assembly.jar
mv: marathon-assembly-0.8.2-RC3.jar: No such file or directory
make: *** [mesos-marathon] Error 1
```

**Comment by mboersma** [comment]
LGTM. I tested with vagrant using `make -C mesos/ deploy` and the documentation in this PR. Nice work @smothiki!

There are undoubtedly rough edges here, but it is a tech preview and we want to get it out there so the Deis community can help us test and refine it. Let's have another maintainer give it a review and then merge.:grinning:

**Comment by carmstrong** [comment]
Code LGTM.

**Comment by smothiki** [comment]
lovely day
